vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,359 @@
+    {
+      co_return;
+    }
+
+    m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    m_writer.WriteBytes(arr);
+
+    co_await m_writer.StoreAsync();
+
+    if (m_pingHandler)
+    {
+      m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  if (m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  try
+  {
+    co_await resume_background();
+
+    co_await resume_on_signal(m_connectPerformed.get());
+
+    if (m_readyState != ReadyState::Open)
+    {
+      co_return;
+    }
+
+    size_t length;
+    std::pair<string, bool> front;
+    bool popped = m_writeQueue.try_pop(front);
+    if (!popped)
+    {
+      throw hresult_error(E_FAIL, L"Could not retrieve outgoing message.");
+    }
+
+    auto [message, isBinary] = std::move(front);
+
+    if (isBinary)
+    {
+      m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      //TODO: Use char_t instead of uint8_t?
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()),
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
+      m_writer.WriteBytes(arr);
+    }
+
+    co_await m_writer.StoreAsync();
+
+    if (m_writeHandler)
+    {
+      m_writeHandler(length);
+    }
+
+    if (!m_writeQueue.empty())
+    {
+      PerformWrite();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformClose()
+{
+  co_await resume_background();
+
+  co_await resume_on_signal(m_connectPerformed.get());
+
+  m_socket.Close(static_cast<uint16_t>(m_closeCode), Utf8ToUtf16(m_closeReason));
+
+  m_readyState = ReadyState::Closed;
+
+  co_return;
+}
+
+void WinRTWebSocketResource::OnMessageReceived(IWebSocket const& sender, MessageWebSocketMessageReceivedEventArgs const& args)
+{
+  try
+  {
+    string response;
+    DataReader reader = args.GetDataReader();
+    auto len = reader.UnconsumedBufferLength();
+    if (args.MessageType() == SocketMessageType::Utf8)
+    {
+      reader.UnicodeEncoding(UnicodeEncoding::Utf8);
+      vector<uint8_t> data(len);
+      reader.ReadBytes(data);
+
+      response = string(CheckedReinterpretCast<char*>(data.data()), data.size());
```

Comment:
>string [](start = 17, length = 6)

Similar to another comment, using uint8_t in one place but "char" in another place is very confusing. We should really try to stick to one type. 
For example, here you have allocation of "vector" in line 224 and then "string" in line 227.
Instead, you should be able to create std::string object with preallocated buffer. Then pass it into ReadBytes. So you can skip creation of "vectror<uint8_t>" altogether.

Reply:
I agree. Though, this is exactly what the UWP team implemented.
I'll consult with them to see if there was a valid reason for this discrepancy.
If there is, I'll add comments. Else, I'll move to `char`.